src/main/scala/chisel3/tester/legacy/backends/vcs/IntervalTest.scala

```diff
@@ -0,0 +1,96 @@
+// See README.md for license details.
+
+package chisel3.tests
+
+// See README.md for license details.
+
+import chisel3._
+import chisel3.experimental._
+import chisel3.internal.firrtl.IntervalRange
+import chisel3.tester._
+import chisel3.tester.experimental.TestOptionBuilder._
+import chisel3.tester.internal.TreadleBackendAnnotation
+import logger.LazyLogging
+import org.scalatest._
+
+/** Shows several usages of Interval working converting larger ranges to smaller ones
+  * To see tabular results useful for debugging add the following line
+  *   {{{
+  *       _root_.logger.Logger.setLevel("chisel3.tests.IntervalTest", _root_.logger.LogLevel.Debug)
+  *   }}}
+  * to the line after {{{class IntervalTest}}}
+  */
+
+class HasIntervalWithBinaryPoint(inputRange: IntervalRange, outputRange: IntervalRange) extends Module {
+  val io = IO(new Bundle {
+    val input = Input(Interval(inputRange))
+
+    val out1 = Output(Interval(inputRange))
+    val out1Clip = Output(Interval(outputRange))
+    val out1Squeeze = Output(Interval(outputRange))
+    val out1Wrap = Output(Interval(outputRange))
+  })
+
+  io.out1 := io.input
+  io.out1Clip := io.input.clip(io.out1Clip)
+  io.out1Squeeze := io.input.squeeze(io.out1Squeeze)
+  io.out1Wrap := io.input.wrap(io.out1Wrap)
+}
+
+class IntervalTest extends FreeSpec with ChiselScalatestTester with LazyLogging with Matchers {
+
+  def intervalTestImpl(inputRange: IntervalRange, outputRange: IntervalRange): Unit = {
+
+    def wrap(b: BigDecimal): BigDecimal = {
+      val minValue: BigDecimal = outputRange.getLowestPossibleValue.get
+      val maxValue: BigDecimal = outputRange.getHighestPossibleValue.get
+      val delta:    BigDecimal = inputRange.increment.get
+      val lowToHigh = (maxValue - minValue) + delta
+      if (b < minValue) {
+        lowToHigh + b
+      } else if (b > maxValue) {
+        -lowToHigh + b
+      } else {
+        b
+      }
+    }
+
+    val inputBinaryPoint = inputRange.binaryPoint.get
+    val outputBinaryPoint = outputRange.binaryPoint.get
+
+    test(new HasIntervalWithBinaryPoint(inputRange, outputRange)).withAnnotations(Seq(TreadleBackendAnnotation)) { c =>
+      logger.debug(f"${"in"}%10s   " + f"${"Clip"}%10s   " + f"${"Squeeze"}%10s   " + f"${"Wrap"}%10s")
+
+      for (i <- inputRange.getPossibleValues) {
+        c.io.input.poke(i.I(binaryPoint = inputBinaryPoint.BP))
+
+        logger.debug(
```

Comment:
Having worked on this stuff multiple times now, it can be really tricky to figure out what's wrong with things particularly with edge cases. When I've written the same debugging code multiple times. I sometimes want to keep it so I don't have to re-invent it when something changes in the future. I believe logging here provides a modestly intrusive way of doing this. There are potential changes to the firrtl as we further debug intervals that make it exceedingly like that I will be debugging this code once again